query/src/cmd/compaction.rs

```diff
@@ -0,0 +1,255 @@
+use common_planners::OptimizeTablePlan;
+use futures::ready;
+use futures::Stream;
+use tokio_stream::StreamExt;
+use tokio_util::time::delay_queue;
+use tokio_util::time::DelayQueue;
+
+use crate::catalogs::CATALOG_DEFAULT;
+use crate::sessions::QueryContext;
+use crate::sessions::TableContext;
+use crate::storages::Table;
+use crate::Config;
+
+pub async fn compaction(ctx: Arc<QueryContext>, conf: &Config) -> common_exception::Result<()> {
+    let rpc_conf = conf.meta.to_meta_grpc_client_conf();
+    if rpc_conf.local_mode() {
+        return Err(ErrorCode::CompactionError(
+            "Auto compaction is not supported in local mode".to_string(),
+        ));
+    }
+
+    let (key, key_end) = get_start_and_end_of_prefix(PREFIX_TABLE_BY_ID)?;
+    let watch = WatchRequest {
+        key,
+        key_end: Some(key_end),
+        filter_type: FilterType::Update.into(),
+    };
+    // Create a new client of metasrv.
+    let client = MetaGrpcClient::try_new(&rpc_conf)?;
+
+    // Watch the table meta update.
+    let mut client_stream = client.request(watch).await?;
+    let task = CompactionQueue::new(Duration::from_secs(3600));
+    let (mut delay_queue_insert, mut delay_queue_read) = task.create_channel(ctx.clone());
+    ctx.get_storage_runtime().spawn(async move {
+        while let Ok(Some(resp)) = client_stream.message().await {
+            if let Some(event) = resp.event {
+                if event.prev.is_none() || event.current.is_none() {
+                    continue;
+                }
+
+                let key = event.key;
+                let id = TableId::from_key(&key).unwrap();
+
+                let current_seqv = event.current.unwrap();
+                let seq = current_seqv.seq;
+                let meta: TableMeta = deserialize_struct(&current_seqv.data).unwrap();
+                if meta.drop_on.is_some() || meta.default_cluster_key.is_some() {
+                    continue;
+                }
+
+                let ident = TableIdent::new(id.table_id, seq);
+                let table_info = TableInfo {
+                    ident,
+                    desc: "".to_owned(),
+                    name: "".to_owned(),
+                    meta,
+                };
+                delay_queue_insert
+                    .insert(id.table_id, table_info)
+                    .await
+                    .unwrap();
+            }
+        }
+    });
+
+    let catalog = ctx.get_catalog(CATALOG_DEFAULT)?;
+    while let Some(info) = delay_queue_read.next().await {
+        let table = catalog.get_table_by_info(&info)?;
+        let ctx = Arc::clone(&ctx);
+        ctx.get_storage_runtime()
+            .spawn(async move { do_compaction(ctx.clone(), table, "".to_owned()).await });
+    }
+
+    Ok(())
+}
+
+pub async fn do_compaction(
+    ctx: Arc<QueryContext>,
+    table: Arc<dyn Table>,
+    database: String,
+) -> common_exception::Result<()> {
+    if !table.auto_compaction() {
+        return Ok(());
```

Comment:
It's better to print some information to user that the table can't auto compaction?